docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+      continued operation of the node over guaranteed delivery of this data, we
+      may eventually run out of disk storage. So here we would need to persist the
+      data _and_ we would need to panic if the sliding window of events fills up.
+   3. Allows for multiple consumers. While desirable from typical web services,
+      Tendermint is a consensus engine and, as such, we should prioritize consensus
+      over providing similar guarantees to a scalable RPC interface (especially
+      since we know this does not scale, and if such a job could just be
+      outsourced). If we have a single consumer, we can easily automatically prune
+      data that was already sent. We could overcome this by limiting the API to a
+      single consumer and single event window.
+
+   Making the requisite modifications to ADR-075 basically converges on the
+   solution outlined in this ADR, except that ADR-075 publishes data via JSON over
+   HTTP while this solution proposes gRPC.
+
+2. We could pick a database that provides real-time notifications of inserts to
+   consumers and just dump the requisite data into that database. Consumers
+   would then need to connect to that database, listen for updates, and then
+   transform/process the data as needed.
+
+   This approach could also inevitably lead to feature sprawl in people wanting
+   us to support multiple different databases. It is also not clear which
+   database would be the best option here.
+
+## Decision
+
+> TODO(thane)
+
+## Detailed Design
+
+### Use Cases
+
+1. Transaction facilitators could provide standalone RPC services that would be
+   capable of receiving transactions and providing confirmations to submitters.
+   These RPC services could be architected to scale independently of the node
+   facilitating consensus.
+2. Block explorers could make use of this data to provide real-time information
+   about the blockchain.
+3. IBC relayer nodes could use this data to filter and store only the data they
+   need to facilitate relaying without putting additional pressure on the
+   consensus engine (until such time that a decision is made on whether to
+   continue providing event data from Tendermint).
+
+### Requirements
+
+1. All of the following data must be published to a single external consumer:
+   1. `FinalizeBlockResponse` data, which includes at the very least:
+      1. Events
+      2. Transaction results
+   2. Committed block data
+
+2. It must be opt-in. In other words, it is off by default and turned on by way
+   of a configuration parameter. When off, it must have negligible (ideally no)
+   impact on system performance.
+
+3. It must not cause back-pressure into consensus.
+
+4. It must not cause unbounded memory growth.
+
+5. Data must be published reliably. If data cannot be published, the node must
+   rather crash in such a way that bringing the node up again will cause it to
+   attempt to republish the unpublished data.
```

Comment:
How do you define "published"? Is it enough that an event is successfully sent onto a Go channel? Written to a socket? Acked at the TCP level?

You can basically pick either reliable delivery _or_ not affecting liveness, but not really both. (Crashing certainly impacts liveness.) And if you pick reliable delivery, you really really want to do that in-process.

Reply:
My assumption here in this document (as I've tried to articulate, but perhaps I didn't do a good job of it) is that users of this interface will be running full nodes (not validators) whose sole purpose is to obtain real-time access to the data this interface exposes.

These consumers would require reliable delivery of the data. That is, if the node cannot deliver the data to the companion, the node must fail.

Think of it almost like a storage device for the node. If the node cannot access its storage, it should crash.

Using gRPC, once the node receives a success response from the companion, the node considers the data to be delivered.

And remember, this interface is opt-in. It's off by default.